File: packages/provide-uterm-go/sessionlogger/sessionlogger.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

// Package sessionlogger provides the JSONL session logger for recording BBS
// sessions. Port of provide.uterm.session_logger.
//
// The Logger field accepts any *slog.Logger; wire it to
// provide-telemetry's GetLogger in application code so records flow through
// the shared telemetry pipeline.
package sessionlogger

import (
	"encoding/base64"
	"encoding/json"
	"log/slog"
	"sync"
	"time"

	"github.com/provide-io/provide-uterm/packages/provide-uterm-go/recording"
	"github.com/provide-io/provide-uterm/packages/provide-uterm-go/redaction"
	"github.com/provide-io/provide-uterm/packages/provide-uterm-go/screen"
	"github.com/provide-io/provide-uterm/packages/provide-uterm-go/session"
)

// ControlChannelMode selects whether wire-level control-channel traffic is
// recorded.
type ControlChannelMode string

// Control-channel recording modes.
const (
	// ModeExclude drops wire/control records (the default).
	ModeExclude ControlChannelMode = "exclude"
	// ModeWire records raw wire chunks and decoded control frames.
	ModeWire ControlChannelMode = "wire"
)

// Options configure a SessionLogger.
type Options struct {
	// MaxBytes caps total bytes written; 0 = unlimited.
	MaxBytes int
	// ControlChannelMode defaults to ModeExclude.
	ControlChannelMode ControlChannelMode
	// Redactor, when set, rewrites logged text.
	Redactor redaction.Redactor
	// FlushInterval is the periodic flush cadence; zero selects 5s.
	FlushInterval time.Duration
	// BatchSize triggers a flush when the buffer reaches it; zero selects 100.
	BatchSize int
	// Logger receives quota/flush warnings; nil selects slog.Default().
	Logger *slog.Logger
}

// SessionLogger is an async session recorder over a pluggable
// recording.Store. Each log entry is a JSON object with at minimum
// {"ts": ..., "event": ..., "data": {...}}.
type SessionLogger struct {
	store              recording.Store
	maxBytes           int
	controlChannelMode ControlChannelMode
	redactor           redaction.Redactor
	flushInterval      time.Duration
	batchSize          int
	logger             *slog.Logger

	mu           sync.Mutex
	sessionID    string
	context      map[string]string
	bytesWritten int
	quotaWarned  bool
	buffer       []recording.Event

	flushStop chan struct{}
	flushDone chan struct{}
}

// New creates a SessionLogger over store.
func New(store recording.Store, opts Options) *SessionLogger {
	if opts.ControlChannelMode == "" {
		opts.ControlChannelMode = ModeExclude
	}
	if opts.FlushInterval == 0 {
		opts.FlushInterval = 5 * time.Second
	}
	if opts.BatchSize == 0 {
		opts.BatchSize = 100
	}
	if opts.Logger == nil {
		opts.Logger = slog.Default()
	}
	return &SessionLogger{
		store:              store,
		maxBytes:           opts.MaxBytes,
		controlChannelMode: opts.ControlChannelMode,
		redactor:           opts.Redactor,
		flushInterval:      opts.FlushInterval,
		batchSize:          opts.BatchSize,
		logger:             opts.Logger,
	}
}

// Start begins a recording session.
func (l *SessionLogger) Start(sessionID string) error {
	l.mu.Lock()
	l.sessionID = sessionID
	l.mu.Unlock()

	metadata := map[string]any{"started_at": float64(time.Now().UnixNano()) / 1e9}
	if err := l.store.StartSession(sessionID, metadata); err != nil {
		return err
	}

	meta, err := l.store.RecordingMeta(sessionID)
	if err != nil {
		return err
	}
	l.mu.Lock()
	l.bytesWritten = int(meta.SizeBytes)
	l.flushStop = make(chan struct{})
	l.flushDone = make(chan struct{})
	l.mu.Unlock()

	go l.periodicFlush(l.flushStop, l.flushDone)
	return nil
}

// Stop finalizes the recording session, flushing buffered entries.
func (l *SessionLogger) Stop() error {
	l.mu.Lock()
	stop, done := l.flushStop, l.flushDone
	l.flushStop, l.flushDone = nil, nil
	l.mu.Unlock()
	if stop != nil {
		close(stop)
		<-done
	}

	if err := l.Flush(); err != nil {
		return err
	}
	l.mu.Lock()
	sessionID := l.sessionID
	l.mu.Unlock()
	if sessionID != "" {
		return l.store.EndSession(sessionID)
	}
	return nil
}

// LogSend logs sent keystrokes (redacted, with CP437 wire bytes).
func (l *SessionLogger) LogSend(keys string) error {
	keys = l.redactText(keys)
	payload := screen.EncodeCP437(keys)
	return l.writeEvent("send", map[string]any{
		"keys":      keys,
		"bytes_b64": base64.StdEncoding.EncodeToString(payload),
	})
}

// LogSendMasked logs a credential send without capturing the actual value.
func (l *SessionLogger) LogSendMasked(byteCount int) error {
	return l.writeEvent("send", map[string]any{
		"keys":       "***",
		"bytes_b64":  base64.StdEncoding.EncodeToString([]byte("***")),
		"masked":     true,
		"byte_count": byteCount,
	})
}

// LogScreen logs a screen snapshot with its raw bytes.
func (l *SessionLogger) LogScreen(snapshot session.Snapshot, raw []byte) error {
	rawText := l.redactText(screen.DecodeCP437(raw))
	rawBytes := screen.EncodeCP437(rawText)
	data := l.snapshotData(snapshot)
	data["raw"] = rawText
	data["raw_bytes_b64"] = base64.StdEncoding.EncodeToString(rawBytes)
	return l.writeEvent("read", data)
}

// snapshotData converts a snapshot to the Python dict shape with redacted
// string values.
func (l *SessionLogger) snapshotData(snap session.Snapshot) map[string]any {
	data := map[string]any{
		"screen":             l.redactText(snap.Screen),
		"screen_hash":        snap.ScreenHash,
		"cursor":             map[string]any{"x": snap.Cursor.X, "y": snap.Cursor.Y},
		"cols":               snap.Cols,
		"rows":               snap.Rows,
		"term":               snap.Term,
		"cursor_at_end":      snap.CursorAtEnd,
		"has_trailing_space": snap.HasTrailingSpace,
		"raw_tail":           l.redactText(snap.RawTail),
		"captured_at":        snap.CapturedAt,
	}
	if snap.PromptDetected != nil {
		data["prompt_detected"] = map[string]any{
			"prompt_id":  snap.PromptDetected.PromptID,
			"input_type": snap.PromptDetected.InputType,
			"is_idle":    snap.PromptDetected.IsIdle,
			"kv_data":    l.redactValue(snap.PromptDetected.KVData),
		}
	}
	return data
}

// LogEvent logs an arbitrary named event.
func (l *SessionLogger) LogEvent(event string, data map[string]any) error {
	return l.writeEvent(event, data)
}

// LogWire logs a raw wire chunk when wire-mode recording is enabled.
// direction is "send" or "recv".
func (l *SessionLogger) LogWire(direction, text string) error {
	if l.controlChannelMode != ModeWire {
		return nil
	}
	text = l.redactText(text)
	return l.writeEvent("wire_"+direction, map[string]any{
		"text":      text,
		"bytes_b64": base64.StdEncoding.EncodeToString([]byte(text)),
	})
}

// LogControl logs a decoded control frame when wire-mode recording is
// enabled. direction is "send" or "recv".
func (l *SessionLogger) LogControl(direction string, control map[string]any) error {
	if l.controlChannelMode != ModeWire {
		return nil
	}
	return l.writeEvent("control_"+direction, map[string]any{"control": control})
}

// SetContext sets metadata context attached to subsequent log entries.
func (l *SessionLogger) SetContext(context map[string]string) {
	l.mu.Lock()
	defer l.mu.Unlock()
	l.context = make(map[string]string, len(context))
	for k, v := range context {
		l.context[k] = v
	}
}

// ClearContext clears the metadata context.
func (l *SessionLogger) ClearContext() {
	l.mu.Lock()
	defer l.mu.Unlock()
	l.context = nil
}

// Flush manually flushes buffered log entries.
func (l *SessionLogger) Flush() error {
	l.mu.Lock()
	defer l.mu.Unlock()
	return l.flushLocked()
}

func (l *SessionLogger) writeEvent(event string, data map[string]any) error {
	l.mu.Lock()
	defer l.mu.Unlock()

	if l.maxBytes > 0 && l.bytesWritten >= l.maxBytes {
		if !l.quotaWarned {
			l.quotaWarned = true
			l.logger.Warn("session_logger_quota_reached — further writes suppressed")
		}
		return nil
	}

	record := recording.Event{
		"ts":    float64(time.Now().UnixNano()) / 1e9,
		"event": event,
		"data":  data,
	}
	if l.sessionID != "" {
		record["session_id"] = l.sessionID
	}
	if len(l.context) > 0 {
		ctx := make(map[string]string, len(l.context))
		for k, v := range l.context {
			ctx[k] = v
		}
		record["ctx"] = ctx
	}

	l.buffer = append(l.buffer, record)
	line, err := json.Marshal(record)
	if err != nil {
		return err
	}
	l.bytesWritten += len(line) + 1

	if len(l.buffer) >= l.batchSize {
		return l.flushLocked()
	}
	return nil
}

// flushLocked flushes the buffer while holding l.mu. The buffer is cleared
// only after AppendEvents succeeds, so a failed flush keeps the batch for the
// next attempt.
func (l *SessionLogger) flushLocked() error {
	if len(l.buffer) == 0 || l.sessionID == "" {
		return nil
	}
	batch := make([]recording.Event, len(l.buffer))
	copy(batch, l.buffer)
	if err := l.store.AppendEvents(l.sessionID, batch); err != nil {
		return err
	}
	l.buffer = l.buffer[:0]
	return nil
}

// periodicFlush retries flushes on a fixed cadence until stopped. A transient
// store failure must not kill the flusher: the batch stays buffered and the
// next tick retries it.
func (l *SessionLogger) periodicFlush(stop <-chan struct{}, done chan<- struct{}) {
	defer close(done)
	ticker := time.NewTicker(l.flushInterval)
	defer ticker.Stop()
	for {
		select {
		case <-stop:
			return
		case <-ticker.C:
			if err := l.Flush(); err != nil {
				l.mu.Lock()
				sessionID := l.sessionID
				l.mu.Unlock()
				l.logger.Warn("session_logger_periodic_flush_failed",
					"session_id", sessionID, "error", err.Error())
			}
		}
	}
}

func (l *SessionLogger) redactText(value string) string {
	return redaction.RedactText(value, l.redactor)
}

func (l *SessionLogger) redactValue(value any) any {
	switch v := value.(type) {
	case string:
		return l.redactText(v)
	case []any:
		out := make([]any, len(v))
		for i, item := range v {
			out[i] = l.redactValue(item)
		}
		return out
	case map[string]any:
		out := make(map[string]any, len(v))
		for k, item := range v {
			out[k] = l.redactValue(item)
		}
		return out
	default:
		return value
	}
}

File: packages/provide-uterm-ts/src/session-logger/session-logger.ts
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

/**
 * JSONL session logger for recording terminal sessions.
 *
 * Each entry is one JSON object carrying at minimum `ts`, `event` and `data`.
 * Entries are buffered and flushed in batches, or on a timer, so a chatty
 * session does not turn into one store round-trip per keystroke.
 *
 * Port of the Python module `provide.uterm.session_logger` and the Go package
 * `sessionlogger`.
 */

import type { RecordingEvent, RecordingStore } from "../recording/index.ts";
import { type Redactor, redactText } from "../redaction/index.ts";
import { decodeCp437, encodeCp437 } from "../screen/index.ts";
import { getLogger, type Logger } from "../telemetry/index.ts";

/** Whether control-channel traffic is recorded. */
export type ControlChannelMode = "exclude" | "wire";

/** Direction of a wire or control entry. */
export type Direction = "send" | "recv";

/** Construction options for {@link SessionLogger}. */
export interface SessionLoggerOptions {
  /** Byte budget for the session. Zero means unlimited. */
  maxBytes?: number;
  /** Whether to record raw wire and decoded control frames. */
  controlChannelMode?: ControlChannelMode;
  /** Applied to every logged string before it is written. */
  redactor?: Redactor;
  /** Seconds between periodic flushes. */
  flushIntervalS?: number;
  /** Entries buffered before a flush is forced. */
  batchSize?: number;
  /** Logger for the recorder's own diagnostics. */
  logger?: Logger;
}

/** Base64-encode bytes the way the reference does. */
function toBase64(data: Uint8Array): string {
  return Buffer.from(data).toString("base64");
}

/** Async session recorder writing through a pluggable store. */
export class SessionLogger {
  readonly #store: RecordingStore;
  readonly #maxBytes: number;
  readonly #controlChannelMode: ControlChannelMode;
  readonly #redactor: Redactor | undefined;
  readonly #flushIntervalMs: number;
  readonly #batchSize: number;
  readonly #logger: Logger;

  #sessionId: string | undefined;
  #context: Record<string, string> = {};
  #bytesWritten = 0;
  #quotaWarned = false;
  #buffer: RecordingEvent[] = [];
  #timer: ReturnType<typeof setInterval> | undefined;
  /** Serialises writes so a flush cannot interleave with an append. */
  #chain: Promise<void> = Promise.resolve();

  constructor(store: RecordingStore, options: SessionLoggerOptions = {}) {
    this.#store = store;
    this.#maxBytes = options.maxBytes ?? 0;
    this.#controlChannelMode = options.controlChannelMode ?? "exclude";
    this.#redactor = options.redactor;
    this.#flushIntervalMs = (options.flushIntervalS ?? 5) * 1000;
    this.#batchSize = options.batchSize ?? 100;
    this.#logger = options.logger ?? getLogger("provide.uterm.session_logger");
  }

  /**
   * Run `work` after everything already queued.
   *
   * The reference holds an async lock across every write; this chains them
   * instead, which gives the same ordering guarantee without a lock object.
   */
  #serialise<T>(work: () => Promise<T>): Promise<T> {
    const result = this.#chain.then(work);
    // Swallow on the chain only: the caller still sees the rejection, but one
    // failed write must not poison every write after it.
    this.#chain = result.then(
      () => undefined,
      () => undefined,
    );
    return result;
  }

  /** Begin a recording session. */
  async start(sessionId: string): Promise<void> {
    this.#sessionId = sessionId;
    await this.#store.startSession(sessionId, { started_at: Date.now() / 1000 });
    const meta = await this.#store.recordingMeta(sessionId);
    const size = Number(meta.size_bytes);
    this.#bytesWritten = Number.isFinite(size) ? size : 0;
    this.#timer = setInterval(() => {
      void this.flush().catch((error: unknown) => {
        // A transient store failure must not stop the periodic flusher: the
        // batch stays buffered, so the next tick retries it. Logged at warn
        // rather than error because it is expected and retried.
        this.#logger.warn(
          { event: "session_logger_periodic_flush_failed", session_id: this.#sessionId, error: String(error) },
          "periodic flush failed",
        );
      });
    }, this.#flushIntervalMs);
    // A recorder must never hold the process open on its own.
    this.#timer.unref?.();
  }

  /** Finalise the recording session. */
  async stop(): Promise<void> {
    if (this.#timer !== undefined) {
      clearInterval(this.#timer);
      this.#timer = undefined;
    }
    await this.flush();
    if (this.#sessionId !== undefined) {
      await this.#store.endSession(this.#sessionId);
    }
  }

  /** Log sent keystrokes. */
  async logSend(keys: string): Promise<void> {
    const redacted = this.#redact(keys);
    await this.#write("send", { keys: redacted, bytes_b64: toBase64(encodeCp437(redacted)) });
  }

  /** Log a credential send without capturing the value. */
  async logSendMasked(byteCount: number): Promise<void> {
    await this.#write("send", {
      keys: "***",
      bytes_b64: toBase64(new Uint8Array(Buffer.from("***", "ascii"))),
      masked: true,
      byte_count: byteCount,
    });
  }

  /** Log a screen snapshot alongside the raw bytes that produced it. */
  async logScreen(snapshot: Record<string, unknown>, raw: Uint8Array): Promise<void> {
    const rawText = this.#redact(decodeCp437(raw));
    await this.#write("read", {
      ...(this.#redactValue(snapshot) as Record<string, unknown>),
      raw: rawText,
      raw_bytes_b64: toBase64(encodeCp437(rawText)),
    });
  }

  /** Log an arbitrary named event. */
  async logEvent(event: string, data: Record<string, unknown>): Promise<void> {
    await this.#write(event, data);
  }

  /** Log a raw wire chunk, when wire-mode recording is enabled. */
  async logWire(direction: Direction, text: string): Promise<void> {
    if (this.#controlChannelMode !== "wire") {
      return;
    }
    const redacted = this.#redact(text);
    await this.#write(`wire_${direction}`, {
      text: redacted,
      bytes_b64: toBase64(new Uint8Array(Buffer.from(redacted, "utf-8"))),
    });
  }

  /** Log a decoded control frame, when wire-mode recording is enabled. */
  async logControl(direction: Direction, control: Record<string, unknown>): Promise<void> {
    if (this.#controlChannelMode !== "wire") {
      return;
    }
    await this.#write(`control_${direction}`, { control });
  }

  /**
   * Set metadata attached to subsequent entries.
   *
   * Values are stringified, matching the reference, so a numeric context
   * value arrives on the wire as text rather than varying by call site.
   */
  setContext(context: Record<string, unknown>): void {
    this.#context = Object.fromEntries(Object.entries(context).map(([key, value]) => [String(key), String(value)]));
  }

  /** Clear the metadata context. */
  clearContext(): void {
    this.#context = {};
  }

  /** Flush buffered entries to the store. */
  flush(): Promise<void> {
    return this.#serialise(() => this.#flushUnlocked());
  }

  /**
   * Write the buffer to the store.
   *
   * The buffer is cleared only after the store accepts the batch: if the
   * store throws, the batch stays buffered for the next attempt rather than
   * being lost.
   */
  async #flushUnlocked(): Promise<void> {
    if (this.#buffer.length === 0 || this.#sessionId === undefined) {
      return;
    }
    const batch = [...this.#buffer];
    await this.#store.appendEvents(this.#sessionId, batch);
    this.#buffer = this.#buffer.slice(batch.length);
  }

  /** Buffer one entry, flushing when the batch is full. */
  #write(event: string, data: Record<string, unknown>): Promise<void> {
    return this.#serialise(async () => {
      if (this.#maxBytes > 0 && this.#bytesWritten >= this.#maxBytes) {
        if (!this.#quotaWarned) {
          this.#quotaWarned = true;
          this.#logger.warn(
            { event: "session_logger_quota_reached", session_id: this.#sessionId },
            "further writes suppressed",
          );
        }
        return;
      }

      const record: RecordingEvent = { ts: Date.now() / 1000, event, data };
      if (this.#sessionId !== undefined) {
        record.session_id = this.#sessionId;
      }
      if (Object.keys(this.#context).length > 0) {
        record.ctx = { ...this.#context };
      }

      this.#buffer.push(record);
      this.#bytesWritten += JSON.stringify(record).length + 1;

      if (this.#buffer.length >= this.#batchSize) {
        await this.#flushUnlocked();
      }
    });
  }

  /** Apply the configured redactor to a string. */
  #redact(value: string): string {
    return redactText(value, this.#redactor);
  }

  /** Apply the redactor to every string inside a value, recursively. */
  #redactValue(value: unknown): unknown {
    if (typeof value === "string") {
      return this.#redact(value);
    }
    if (Array.isArray(value)) {
      return value.map((item) => this.#redactValue(item));
    }
    if (typeof value === "object" && value !== null) {
      return Object.fromEntries(Object.entries(value).map(([key, item]) => [key, this.#redactValue(item)]));
    }
    return value;
  }
}
